### DISCUSSION TITLE
Testing optional keys only

### DISCUSSION THREAD

**Discussion by mahadi** [crdoconnor/strictyaml#203] [Q&A] [answered]
As a part of my application i am testing the schema validation. For a specific section of my configuration i have a Map which contains an optional username and password (and other keys).

I want to test that both combinations wont raise an error during the validation.
To be able to test only the relevant parts of my schema i only extracted the username and password schema part and provided an empty string, simulating that none of the keys has been provided.

I only want to test the optional keys because i already have tests for the other keys, so i dont want to include any of them again.

Shouldnt the following test pass?

``` python
from strictyaml import Map, Str, Optional, load
from ensure import Ensure
schema = Map({Optional("username"): Str(), Optional("password"): Str()})
Ensure(load('', schema)).equals({})
```
It raises:

```
strictyaml.exceptions.YAMLValidationError: when expecting a mapping
found a blank string
  in "<unicode string>", line 1, column 1:
    ''
     ^ (line: 1)
```

**Comment by crdoconnor** [comment, marked as answer]
No, it shouldn't. The Map validator implies that there should be *at least one* key and value, even though both individually are optional. If you wanted it to be possible to have neither, you would need

```EmptyDict() | Map({Optional("username"): Str(), Optional("password"): Str()}).```

**Reply by mahadi** [reply]
Thank you for the explanation, it works as you proposed.
